### PR TITLE
Upgrade pgx dependencies to 0.2.0

### PR DESCRIPTION
**Pull request by JamesGuthrie**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,12 +14,13 @@ lto = true
...
 serde = { version = "1.0", features = ["derive"] }
-pgx = {git="https://github.com/JLockerman/pgx.git", branch="guardless"}
-pgx-macros = {git="https://github.com/JLockerman/pgx.git", branch="guardless"}
+pgx = "0.2.0"
+pgx-macros = "0.2.0"
 
 [dev-dependencies]
-pgx-tests = {git="https://github.com/JLockerman/pgx.git", branch="guardless"}
\ No newline at end of file
+pgx-tests = "0.2.0"
```
Comment (JLockerman) [inline Cargo.toml]: might want to up this to `0.2.1` or do a `cargo update`

Comment (JLockerman) [reply Cargo.toml]: In fact, there's going to be a `0.2.2` soon...

Comment (JLockerman) [reply Cargo.toml]: (they're mostly schema generator errors, so the update can wait until that PR)

File: src/lib.rs [hunk 2]
```diff
@@ -586,5 +586,6 @@ CREATE AGGREGATE vector_selector(
     deserialfunc = vector_selector_deserialize,
     parallel = safe
 );
-"#
+"#,
+    name = "create_aggregate_vector_selector"
```
Comment (JLockerman) [inline src/lib.rs:590]: oh, I forgot about these!
It won't really matter until you start using the pgx schema generator, but you should add
```suggestion
    name = "create_aggregate_vector_selector",
    requires = [vector_selector_transition, vector_selector_final, vector_selector_combine, vector_selector_serialize, vector_selector_deserialize],
```
so the generator outputs the SQL in the correct order.

File: src/type_builder.rs [hunk 3]
```diff
@@ -30,20 +30,24 @@ macro_rules! do_serialize {
             unsafe {
                 ::pgx::set_varsize(bytes.as_mut_ptr() as *mut _, bytes.len() as i32);
             }
-            bytes.leak().as_mut_ptr() as pg_sys::Datum
+            bytea::from(bytes.leak().as_mut_ptr() as pg_sys::Datum)
```
Comment (JLockerman) [inline src/type_builder.rs:33]: I totally gave you the wrong version of this file, didn't I 😞 
The correct version is
```rust
#[repr(u8)]
pub enum SerializationType {
    Default = 1,
}

#[macro_export]
macro_rules! do_serialize {
    ($state: ident) => {
        {
            $crate::do_serialize!($state, version: 1)
        }
    };
    ($state: ident, version: $version: expr) => {
        {
            use $crate::type_builder::SerializationType;
            use std::io::{Cursor, Write};
            use std::convert::TryInto;

            let state = &*$state;
            let serialized_size = bincode::serialized_size(state)
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            let our_size = serialized_size + 2; // size of serialized data + our version flags
            let allocated_size = our_size + 4; // size of our data + the varlena header
            let allocated_size = allocated_size.try_into()
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            // valena tyes have a maximum size
            if allocated_size > 0x3FFFFFFF {
                pgx::error!("size {} bytes is to large", allocated_size)
            }

            let bytes: &mut [u8] = unsafe {
                let bytes = pgx::pg_sys::palloc0(allocated_size);
                std::slice::from_raw_parts_mut(bytes.cast(), allocated_size)
            };
            let mut writer = Cursor::new(bytes);
            // varlena header space
            let varsize = [0; 4];
            writer.write_all(&varsize)
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            // type version
            writer.write_all(&[$version])
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            // serialization version; 1 for bincode is currently the only option
            writer.write_all(&[SerializationType::Default as u8])
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            bincode::serialize_into(&mut writer, state)
                .unwrap_or_else(|e| pgx::error!("serialization error {}", e));
            unsafe {
                let len = writer.position().try_into().expect("serialized size too large");
                ::pgx::set_varsize(writer.get_mut().as_mut_ptr() as *mut _, len);
            }
            writer.into_inner().as_mut_ptr() as pg_sys::Datum
        }
    };
}
#[macro_export]
macro_rules! do_deserialize {
    ($bytes: ident, $t: ty) => {
        {
            use $crate::type_builder::SerializationType;

            let state: $t = unsafe {
                let detoasted = pg_sys::pg_detoast_datum_packed($bytes as *mut _);
                let len = pgx::varsize_any_exhdr(detoasted);
                let data = pgx::vardata_any(detoasted);
                let bytes = slice::from_raw_parts(data as *mut u8, len);
                if bytes.len() < 1 {
                    pgx::error!("deserialization error, no bytes")
                }
                if bytes[0] != 1 {
                    pgx::error!("deserialization error, invalid serialization version {}", bytes[0])
                }
                if bytes[1] != SerializationType::Default as u8 {
                    pgx::error!("deserialization error, invalid serialization type {}", bytes[1])
                }
                bincode::deserialize(&bytes[2..]).unwrap_or_else(|e|
                    pgx::error!("deserialization error {}", e))
            };
            state.into()
        }
    };
}
```
from https://github.com/timescale/timescaledb-toolkit/blob/3753078a24192101381ed4696696c48f36b6209c/extension/src/type_builder.rs#L310-L391

File: Cargo.toml (hunk 1 above)
Comment (JamesGuthrie) [reply Cargo.toml]: Yes, I would like to update this in a future change.

File: src/lib.rs (hunk 2 above)
Comment (JamesGuthrie) [reply src/lib.rs:590]: Yes, thanks. I will address this when I get to the schema generator.

File: src/type_builder.rs (hunk 3 above)
Comment (JamesGuthrie) [reply src/type_builder.rs]: Thanks!

**Review by JLockerman — APPROVED** [review]

**Review by cevian — APPROVED** [review]
LGTM. One question about Cargo.Lock but don't feel strongly (we should just do whatever is best practice for Rust)

**Comment by cevian** [comment]
@JamesGuthrie feel free to merge when ready

File: Cargo.lock [hunk 4]
```diff
@@ -1,5 +1,16 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
+version = 3
+
+[[package]]
```
[the remark this answers is not in the document]
Comment (JamesGuthrie) [reply Cargo.lock:5]: The `Cargo.lock` file is managed by Cargo and used to ensure reproducibile builds (using the same dependency versions).

The [Cargo book](https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html) recommends checking `Cargo.lock` in to the repo for binaries and system libraries:

> If you’re building a non-end product, such as a rust library that other rust packages will depend on, put Cargo.lock in your .gitignore. If you’re building an end product, which are executable like command-line tool or an application, or a system library with crate-type of staticlib or cdylib, check Cargo.lock into git. If you're curious about why that is, see "Why do binaries have Cargo.lock in version control, but not libraries?" in the FAQ.

[the remark this answers is not in the document]
Comment (JamesGuthrie) [reply Cargo.lock:5]: Conceptually I guess it's similar to `go.mod`

**Pull request merged**